zcash_client_sqlite/src/wallet.rs

```diff
@@ -375,40 +379,63 @@ pub(crate) fn add_account<P: consensus::Parameters>(
+                account_kind, hd_seed_fingerprint, hd_account_index,
+                ufvk, uivk,
+                orchard_fvk_item_cache, sapling_fvk_item_cache, p2pkh_fvk_item_cache,
+                birthday_height, birthday_sapling_tree_size, birthday_orchard_tree_size,
+                recover_until_height
+            )
+            VALUES (
+                :account_kind, :hd_seed_fingerprint, :hd_account_index,
+                :ufvk, :uivk,
+                :orchard_fvk_item_cache, :sapling_fvk_item_cache, :p2pkh_fvk_item_cache,
+                :birthday_height, :birthday_sapling_tree_size, :birthday_orchard_tree_size,
+                :recover_until_height
+            )
+            RETURNING id;
+            "#,
+            named_params![
+                ":account_kind": account_kind_code(kind),
+                ":hd_seed_fingerprint": hd_seed_fingerprint.as_ref().map(|fp| fp.to_bytes()),
+                ":hd_account_index": hd_account_index.map(u32::from),
+                ":ufvk": ufvk_encoded,
+                ":uivk": viewing_key.uivk().encode(params),
+                ":orchard_fvk_item_cache": orchard_item,
+                ":sapling_fvk_item_cache": sapling_item,
+                ":p2pkh_fvk_item_cache": transparent_item,
+                ":birthday_height": u32::from(birthday.height()),
+                ":birthday_sapling_tree_size": birthday_sapling_tree_size,
+                ":birthday_orchard_tree_size": birthday_orchard_tree_size,
+                ":recover_until_height": birthday.recover_until().map(u32::from)
+            ],
+            |row| Ok(AccountId(row.get(0)?)),
         )
-        RETURNING id;
-        "#,
-        named_params![
-            ":account_kind": account_kind_code(kind),
-            ":hd_seed_fingerprint": hd_seed_fingerprint.as_ref().map(|fp| fp.to_bytes()),
-            ":hd_account_index": hd_account_index.map(u32::from),
-            ":ufvk": viewing_key.ufvk().map(|ufvk| ufvk.encode(params)),
-            ":uivk": viewing_key.uivk().encode(params),
-            ":orchard_fvk_item_cache": orchard_item,
-            ":sapling_fvk_item_cache": sapling_item,
-            ":p2pkh_fvk_item_cache": transparent_item,
-            ":birthday_height": u32::from(birthday.height()),
-            ":birthday_sapling_tree_size": birthday_sapling_tree_size,
-            ":birthday_orchard_tree_size": birthday_orchard_tree_size,
-            ":recover_until_height": birthday.recover_until().map(u32::from)
-        ],
-        |row| Ok(AccountId(row.get(0)?)),
-    )?;
+        .map_err(|e| match e {
+            rusqlite::Error::SqliteFailure(f, s)
+                if f.code == rusqlite::ErrorCode::ConstraintViolation =>
+            {
+                // An account conflict occurred.
+                // Make a best effort to determine the AccountId of the pre-existing row
+                // and provide that to our caller.
+                if s.clone().is_some_and(|s| s.contains(".ufvk")) {
```

Comment:
This check is incompatible with test or regtest keys; instead, one would want to take the prefix of the serialized key up to the character `1` . Also, this depends upon the stability of the text encoding of rusqlite error messages, which I'm not comfortable with; instead, simply query for the ufvk and if no results are returned, then return the original error.

Reply:
I don't know how this fails with test or regtest keys. The substring I'm testing for here is for the name of the column in the table which violated the constraint. So as long as the error mentions the column, it should continue to work. 
But if you want me to drop that check and just go straight to the next level check that queries the db to see if there is a conflicting ufvk (which I already do next), I can drop this first check. It was just an attempt to avoid another db hit when possible.

Reply:
Yeah, please go ahead and drop the check; this should be a rare error case anyway, and I'd prefer not to rely on error strings that we don't control.

Reply:
Done.